lib/auth/bot.go

```diff
@@ -0,0 +1,286 @@
+	if err != nil {
+		err = trace.WrapWithMessage(err, "could not fetch expected bot user %s", resourceName)
+	} else {
+		label, ok := user.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete user %s that is missing label %s", resourceName, types.BotLabel)
+		} else if label != botName {
+			err = trace.Errorf("will not delete user %s with mismatched label %s = %s", resourceName, types.BotLabel, label)
+		} else {
+			err = s.DeleteUser(ctx, resourceName)
+		}
+	}
+
+	return err
+}
+
+// deleteBotRole removes an existing bot role, ensuring that it has bot labels
+// matching the bot before deleting anything.
+func (s *Server) deleteBotRole(ctx context.Context, botName, resourceName string) error {
+	role, err := s.GetRole(ctx, resourceName)
+	if err != nil {
+		err = trace.WrapWithMessage(err, "could not fetch expected bot role %s", resourceName)
+	} else {
+		label, ok := role.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete role %s that is missing label %s", resourceName, types.BotLabel)
+		} else if label != botName {
+			err = trace.Errorf("will not delete role %s with mismatched label %s = %s", resourceName, types.BotLabel, label)
+		} else {
+			err = s.DeleteRole(ctx, resourceName)
+		}
+	}
+
+	return err
+}
+
+func (s *Server) deleteBot(ctx context.Context, botName string) error {
+	// TODO:
+	// remove any locks for the bot's impersonator role?
+	// remove the bot's user
+	resourceName := botResourceName(botName)
+
+	userErr := s.deleteBotUser(ctx, botName, resourceName)
+	roleErr := s.deleteBotRole(ctx, botName, resourceName)
+	return trace.NewAggregate(userErr, roleErr)
+}
+
+// getBotUsers fetches all Users with the BotLabel field set. Users are fetched
+// without secrets.
+func (s *Server) getBotUsers(ctx context.Context) ([]types.User, error) {
+	var botUsers []types.User
+
+	users, err := s.GetUsers(false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, user := range users {
+		if _, ok := user.GetMetadata().Labels[types.BotLabel]; ok {
+			botUsers = append(botUsers, user)
+		}
+	}
+
+	return botUsers, nil
+}
+
+// CreateBotJoinToken creates a new joining token for bots.
+func (s *Server) CreateBotJoinToken(ctx context.Context, req CreateUserTokenRequest) (types.UserToken, error) {
+	err := req.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if req.Type != UserTokenTypeBot {
+		return nil, trace.BadParameter("invalid bot token request type")
+	}
+
+	_, err = s.GetUser(req.Name, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// TODO: ensure that the user is a bot user?
+	token, err := s.newUserToken(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// TODO: deleteUserTokens?
+	token, err = s.Identity.CreateUserToken(ctx, token)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// TODO: audit log event. partially implemented, may need events.proto and
```

Comment:
For things like this maybe you want to remove the TODO and just track this in a separate issue.

Reply:
I've created https://github.com/gravitational/teleport/issues/10477 to track missing audit events